flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/functions/SqlDateTimeUtils.java

```diff
@@ -205,6 +214,36 @@ public static long toTimestamp(Decimal v) {
 	// --------------------------------------------------------------------------------------------
 	// String --> Timestamp conversion
 	// --------------------------------------------------------------------------------------------
+	public static SqlTimestamp toSqlTimestamp(String dateStr) {
+		int length = dateStr.length();
+		String format;
+		if (length == 10) {
+			format = DATE_FORMAT_STRING;
+		} else if (length >= 21 && length <= 29) {
+			format = DEFAULT_DATETIME_FORMATS[length - 20];
+		} else {
+			// otherwise fall back to second's precision
+			format = DEFAULT_DATETIME_FORMATS[0];
+		}
+		return toSqlTimestamp(dateStr, format);
+	}
+
+	public static SqlTimestamp toSqlTimestamp(String dateStr, String format) {
+		DateTimeFormatter formatter = DateTimeFormatter.ofPattern(format);
```

Comment:
`DateTimeFormatter.ofPattern` is very expensive performance, please add cache like previous `FORMATTER_CACHE`.